File: dev-portal/python/sdk/postman_sdk/client/postman_client.py
"""
Postman API Client

Main SDK entry point with all API methods
"""

from typing import Any
from urllib.parse import quote

from postman_sdk.client.http_client import HttpClient, get_error_message
from postman_sdk.types import (
    PostmanClientConfig,
    CurrentUser,
    Workspace,
    WorkspaceRole,
    WorkspaceRoleId,
    CreateWorkspaceRequest,
    CreateWorkspaceResult,
    AddAdminResult,
    Collection,
    CollectionDetails,
    ForkResult,
    Environment,
    EnvironmentDetails,
    EnvironmentVariable,
    CreateEnvironmentResult,
    UpdateEnvironmentResult,
    MockServer,
    CreateMockRequest,
    CreateMockResult,
    Spec,
    SpecFile,
    SpecFileWithContent,
    SpecType,
    CreateSpecFile,
    CreateSpecResult,
    DeleteSpecResult,
    InvitePartnerResult,
    RemovePartnerResult,
)


class PostmanClient:
    """Postman API Client SDK"""

    def __init__(self, api_key: str, **kwargs: Any):
        config = PostmanClientConfig(api_key=api_key, **kwargs)
        self._http = HttpClient(config)

    async def __aenter__(self) -> "PostmanClient":
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

    async def close(self) -> None:
        """Close the client"""
        await self._http.close()

    # =========================================================================
    # USER / AUTHENTICATION
    # =========================================================================

    async def validate_api_key(self) -> dict[str, Any]:
        """Validate API key and get current user info"""
        try:
            response = await self._http.get("/me")
            user_data = response.get("user", {})
            return {"valid": True, "user": CurrentUser(**user_data)}
        except Exception as e:
            return {"valid": False, "error": get_error_message(e)}

    async def get_current_user(self) -> CurrentUser:
        """Get current user info"""
        response = await self._http.get("/me")
        return CurrentUser(**response.get("user", {}))

    # =========================================================================
    # WORKSPACES
    # =========================================================================

    async def get_workspace(self, workspace_id: str) -> Workspace | None:
        """Get workspace details"""
        try:
            response = await self._http.get(f"/workspaces/{workspace_id}")
            return Workspace(**response.get("workspace", {}))
        except Exception:
            return None

    async def create_workspace(self, request: CreateWorkspaceRequest) -> CreateWorkspaceResult:
        """Create a new workspace"""
        try:
            response = await self._http.post(
                "/workspaces",
                {
                    "workspace": {
                        "name": request.name,
                        "type": request.type,
                        "description": request.description or f"Created via SDK",
                    }
                },
            )
            return CreateWorkspaceResult(
                success=True,
                workspace=Workspace(**response.get("workspace", {})),
            )
        except Exception as e:
            return CreateWorkspaceResult(success=False, error=get_error_message(e))

    async def update_workspace(
        self, workspace_id: str, updates: dict[str, Any]
    ) -> dict[str, Any]:
        """Update a workspace via PUT /workspaces/{workspaceId}"""
        try:
            response = await self._http.put(
                f"/workspaces/{workspace_id}", {"workspace": updates}
            )
            return {"success": True, "workspace": response.get("workspace")}
        except Exception as e:
            print(f"Error updating workspace: {e}")
            return {"success": False}

    async def delete_workspace(self, workspace_id: str) -> bool:
        """Delete a workspace"""
        try:
            await self._http.delete(f"/workspaces/{workspace_id}")
            return True
        except Exception:
            return False

    # =========================================================================
    # WORKSPACE ROLES
    # =========================================================================

    async def get_workspace_roles(self, workspace_id: str) -> dict[str, Any]:
        """Get workspace roles"""
        try:
            response = await self._http.get(f"/workspaces/{workspace_id}/roles")
            roles_data = response.get("roles", [])
            roles = [WorkspaceRole(**r) for r in roles_data]
            return {"success": True, "roles": roles}
        except Exception as e:
            return {"success": False, "roles": [], "error": get_error_message(e)}

    async def add_workspace_admin(
        self,
        workspace_id: str,
        user_id: str,
        role_id: str = WorkspaceRoleId.ADMIN,
    ) -> AddAdminResult:
        """Add workspace admin"""
        try:
            response = await self._http.patch(
                f"/workspaces/{workspace_id}/roles",
                {
                    "roles": [
                        {
                            "op": "add",
                            "path": "/user",
                            "value": [{"id": user_id, "role": role_id}],
                        }
                    ]
                },
            )
            roles_data = response.get("roles", [])
            return AddAdminResult(
                success=True,
                roles=[WorkspaceRole(**r) for r in roles_data],
            )
        except Exception as e:
            return AddAdminResult(success=False, error=get_error_message(e))

    # =========================================================================
    # COLLECTIONS
    # =========================================================================

    async def get_collections(self, workspace_id: str) -> list[Collection]:
        """Get all collections in a workspace"""
        try:
            response = await self._http.get(f"/collections?workspace={workspace_id}")
            return [Collection(**c) for c in response.get("collections", [])]
        except Exception:
            return []

    async def get_collection_details(self, collection_uid: str) -> CollectionDetails | None:
        """Get collection details"""
        try:
            response = await self._http.get(f"/collections/{collection_uid}")
            return CollectionDetails(**response.get("collection", {}))
        except Exception:
            return None

    async def fork_collection(
        self,
        collection_uid: str,
        label: str,
        target_workspace_id: str,
    ) -> ForkResult:
        """Fork a collection"""
        try:
            response = await self._http.post(
                f"/collections/fork/{collection_uid}?workspace={target_workspace_id}",
                {"label": label},
            )
            return ForkResult(
                success=True,
                collection=Collection(**response.get("collection", {})),
            )
        except Exception as e:
            return ForkResult(success=False, error=get_error_message(e))

    async def delete_collection(self, collection_uid: str) -> bool:
        """Delete a collection"""
        try:
            await self._http.delete(f"/collections/{collection_uid}")
            return True
        except Exception:
            return False

    async def patch_collection_variables(
        self, collection_uid: str, variables: list[dict[str, Any]]
    ) -> dict[str, Any]:
        """Update a collection's variables via PATCH /collections/{uid}"""
        try:
            response = await self._http.patch(
                f"/collections/{collection_uid}",
                json={"collection": {"variable": variables}},
            )
            return {"success": True, "collection": response.get("collection")}
        except Exception as e:
            return {"success": False, "error": get_error_message(e)}

    # =========================================================================
    # ENVIRONMENTS
    # =========================================================================

    async def get_environments(self, workspace_id: str) -> list[Environment]:
        """Get all environments in a workspace"""
        try:
            response = await self._http.get(f"/environments?workspace={workspace_id}")
            return [Environment(**e) for e in response.get("environments", [])]
        except Exception:
            return []

    async def get_environment_details(self, environment_uid: str) -> EnvironmentDetails | None:
        """Get environment details with variables"""
        try:
            response = await self._http.get(f"/environments/{environment_uid}")
            return EnvironmentDetails(**response.get("environment", {}))
        except Exception:
            return None

    async def create_environment(
        self,
        name: str,
        values: list[EnvironmentVariable],
        workspace_id: str,
    ) -> CreateEnvironmentResult:
        """Create environment"""
        try:
            response = await self._http.post(
                f"/environments?workspace={workspace_id}",
                {
                    "environment": {
                        "name": name,
                        "values": [v.model_dump() for v in values],
                    }
                },
            )
            return CreateEnvironmentResult(
                success=True,
                environment=Environment(**response.get("environment", {})),
            )
        except Exception as e:
            return CreateEnvironmentResult(success=False, error=get_error_message(e))

    async def update_environment(
        self,
        environment_uid: str,
        name: str,
        values: list[EnvironmentVariable],
    ) -> UpdateEnvironmentResult:
        """Update environment (full replace)"""
        try:
            response = await self._http.put(
                f"/environments/{environment_uid}",
                {
                    "environment": {
                        "name": name,
                        "values": [v.model_dump() for v in values],
                    }
                },
            )
            return UpdateEnvironmentResult(
                success=True,
                environment=Environment(**response.get("environment", {})),
            )
        except Exception as e:
            return UpdateEnvironmentResult(success=False, error=get_error_message(e))

    async def delete_environment(self, environment_uid: str) -> bool:
        """Delete environment"""
        try:
            await self._http.delete(f"/environments/{environment_uid}")
            return True
        except Exception:
            return False

    # =========================================================================
    # MOCK SERVERS
    # =========================================================================

    async def get_mocks(self, workspace_id: str) -> list[MockServer]:
        """Get all mocks in a workspace"""
        try:
            response = await self._http.get(f"/mocks?workspace={workspace_id}")
            return [MockServer(**m) for m in response.get("mocks", [])]
        except Exception:
            return []

    async def create_mock(self, request: CreateMockRequest) -> CreateMockResult:
        """Create mock server"""
        try:
            mock_config: dict[str, Any] = {
                "name": request.name,
                "collection": request.collection,
                "private": request.is_private,
            }
            if request.environment:
                mock_config["environment"] = request.environment

            response = await self._http.post(
                f"/mocks?workspace={request.workspace_id}",
                {"mock": mock_config},
            )
            return CreateMockResult(
                success=True,
                mock=MockServer(**response.get("mock", {})),
            )
        except Exception as e:
            return CreateMockResult(success=False, error=get_error_message(e))

    async def delete_mock(self, mock_id: str) -> bool:
        """Delete mock server"""
        try:
            await self._http.delete(f"/mocks/{mock_id}")
            return True
        except Exception:
            return False

    # =========================================================================
    # SPECS
    # =========================================================================

    async def get_specs(self, workspace_id: str) -> list[Spec]:
        """Get all specs in a workspace"""
        try:
            response = await self._http.get(f"/specs?workspaceId={workspace_id}")
            return [Spec(**s) for s in response.get("specs", [])]
        except Exception:
            return []

    async def get_spec_files(self, spec_id: str) -> list[SpecFile]:
        """Get all files in a spec"""
        try:
            response = await self._http.get(f"/specs/{spec_id}/files")
            return [SpecFile(**f) for f in response.get("files", [])]
        except Exception:
            return []

    async def get_spec_file(self, spec_id: str, file_path: str) -> SpecFileWithContent | None:
        """Get a specific spec file with content"""
        try:
            encoded_path = quote(file_path, safe="")
            response = await self._http.get(f"/specs/{spec_id}/files/{encoded_path}")
            return SpecFileWithContent(**response)
        except Exception:
            return None

    async def create_spec(
        self,
        workspace_id: str,
        name: str,
        spec_type: SpecType,
        files: list[CreateSpecFile],
    ) -> CreateSpecResult:
        """Create a spec with files"""
        try:
            response = await self._http.post(
                f"/specs?workspaceId={workspace_id}",
                {
                    "name": name,
                    "type": spec_type,
                    "files": [f.model_dump() for f in files],
                },
            )
            return CreateSpecResult(success=True, spec=Spec(**response))
        except Exception as e:
            return CreateSpecResult(success=False, error=get_error_message(e))

    async def delete_spec(self, spec_id: str) -> DeleteSpecResult:
        """Delete a spec"""
        try:
            await self._http.delete(f"/specs/{spec_id}")
            return DeleteSpecResult(success=True)
        except Exception as e:
            return DeleteSpecResult(success=False, error=get_error_message(e))

    # =========================================================================
    # PARTNER INVITATIONS
    # =========================================================================

    async def invite_partner(
        self,
        workspace_id: str,
        email: str,
        role_id: str = WorkspaceRoleId.PARTNER_EDITOR_AND_LEAD,
    ) -> InvitePartnerResult:
        """Invite partner to workspace"""
        try:
            response = await self._http.post(
                "/invitations",
                {
                    "action": "invite_partner",
                    "targetEntity": "workspace",
                    "targetEntityId": workspace_id,
                    "roleId": role_id,
                    "target": {"emails": [email]},
                },
            )
            results = response.get("results", [])
            result = results[0] if results else {}
            return InvitePartnerResult(
                success=True,
                email=result.get("email", email),
                status=result.get("status"),
                invitation_link=result.get("invitationLink"),
                user_id=result.get("userId"),
                role_display_name=response.get("roleDisplayName"),
            )
        except Exception as e:
            return InvitePartnerResult(success=False, email=email, error=get_error_message(e))

    async def remove_partner(self, workspace_id: str, user_id: str) -> RemovePartnerResult:
        """Remove partner from workspace"""
        try:
            response = await self._http.post(
                "/invitations",
                {
                    "action": "remove_partner",
                    "targetEntity": "workspace",
                    "targetEntityId": workspace_id,
                    "target": {"userIds": [user_id]},
                },
            )
            results = response.get("results", [])
            result = results[0] if results else {}
            return RemovePartnerResult(
                success=True,
                user_id=result.get("userId", user_id),
                status=result.get("status"),
            )
        except Exception as e:
            return RemovePartnerResult(success=False, user_id=user_id, error=get_error_message(e))

File: dev-portal/typescript/sdk/src/client/PostmanClient.ts
/**
 * Postman API Client
 * 
 * Main SDK entry point with all API methods
 */

import { HttpClient, getErrorMessage } from './HttpClient';
import {
  PostmanClientConfig,
  CurrentUser,
  // Workspace types
  Workspace,
  WorkspaceDetails,
  WorkspaceRole,
  WorkspaceRoleId,
  CreateWorkspaceRequest,
  CreateWorkspaceResult,
  AddAdminResult,
  // Collection types
  Collection,
  CollectionDetails,
  ForkResult,
  // Environment types
  Environment,
  EnvironmentDetails,
  EnvironmentVariable,
  CreateEnvironmentResult,
  UpdateEnvironmentResult,
  PatchEnvironmentOperation,
  // Mock types
  MockServer,
  CreateMockRequest,
  CreateMockResult,
  // Spec types
  Spec,
  SpecFile,
  SpecFileWithContent,
  SpecType,
  CreateSpecFile,
  CreateSpecResult,
  DeleteSpecResult,
  // Invitation types
  InvitePartnerResult,
  RemovePartnerResult,
} from '../types';

/**
 * Postman API Client SDK
 */
export class PostmanClient {
  private http: HttpClient;

  constructor(config: PostmanClientConfig) {
    this.http = new HttpClient(config);
  }

  // =========================================================================
  // USER / AUTHENTICATION
  // =========================================================================

  /**
   * Validate API key and get current user info
   */
  async validateApiKey(): Promise<{ valid: boolean; user?: CurrentUser; error?: string }> {
    try {
      const response = await this.http.get<{ user: CurrentUser }>('/me');
      return { valid: true, user: response.user };
    } catch (error) {
      return { valid: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Get current user info
   */
  async getCurrentUser(): Promise<CurrentUser> {
    const response = await this.http.get<{ user: CurrentUser }>('/me');
    return response.user;
  }

  // =========================================================================
  // WORKSPACES
  // =========================================================================

  /**
   * Get workspace details
   */
  async getWorkspace(workspaceId: string): Promise<Workspace | null> {
    try {
      const response = await this.http.get<{ workspace: Workspace }>(`/workspaces/${workspaceId}`);
      return response.workspace;
    } catch {
      return null;
    }
  }

  /**
   * Get workspace details including contents
   */
  async getWorkspaceDetails(workspaceId: string): Promise<WorkspaceDetails | null> {
    try {
      const response = await this.http.get<{ workspace: WorkspaceDetails }>(`/workspaces/${workspaceId}`);
      return response.workspace;
    } catch {
      return null;
    }
  }

  /**
   * Create a new workspace
   */
  async createWorkspace(request: CreateWorkspaceRequest): Promise<CreateWorkspaceResult> {
    try {
      const response = await this.http.post<{ workspace: Workspace }>('/workspaces', {
        workspace: {
          name: request.name,
          type: request.type,
          description: request.description ?? `Created via SDK on ${new Date().toISOString().split('T')[0]}`,
        },
      });
      return { success: true, workspace: response.workspace };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Update a workspace
   */
  async updateWorkspace(
    workspaceId: string,
    updates: Record<string, unknown>,
  ): Promise<{ success: boolean; workspace?: Workspace }> {
    try {
      const response = await this.http.put<{ workspace: Workspace }>(`/workspaces/${workspaceId}`, {
        workspace: updates,
      });
      return { success: true, workspace: response.workspace };
    } catch (error) {
      console.error('Error updating workspace:', getErrorMessage(error));
      return { success: false };
    }
  }

  /**
   * Delete a workspace
   */
  async deleteWorkspace(workspaceId: string): Promise<boolean> {
    try {
      await this.http.delete(`/workspaces/${workspaceId}`);
      return true;
    } catch {
      return false;
    }
  }

  // =========================================================================
  // WORKSPACE ROLES
  // =========================================================================

  /**
   * Get workspace roles
   */
  async getWorkspaceRoles(workspaceId: string): Promise<{ success: boolean; roles: WorkspaceRole[]; error?: string }> {
    try {
      const response = await this.http.get<{ roles: WorkspaceRole[] }>(`/workspaces/${workspaceId}/roles`);
      return { success: true, roles: response.roles || [] };
    } catch (error) {
      return { success: false, roles: [], error: getErrorMessage(error) };
    }
  }

  /**
   * Add workspace admin
   */
  async addWorkspaceAdmin(
    workspaceId: string,
    userId: string,
    roleId: string = WorkspaceRoleId.Admin
  ): Promise<AddAdminResult> {
    try {
      const response = await this.http.patch<{ roles: WorkspaceRole[] }>(`/workspaces/${workspaceId}/roles`, {
        roles: [
          {
            op: 'add',
            path: '/user',
            value: [{ id: userId, role: roleId }],
          },
        ],
      });
      return { success: true, roles: response.roles };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Remove workspace user
   */
  async removeWorkspaceUser(
    workspaceId: string,
    userId: string,
    roleId: string
  ): Promise<{ success: boolean; error?: string }> {
    try {
      await this.http.patch(`/workspaces/${workspaceId}/roles`, {
        roles: [
          {
            op: 'remove',
            path: '/user',
            value: [{ id: userId, role: roleId }],
          },
        ],
      });
      return { success: true };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  // =========================================================================
  // COLLECTIONS
  // =========================================================================

  /**
   * Get all collections in a workspace
   */
  async getCollections(workspaceId: string): Promise<Collection[]> {
    try {
      const response = await this.http.get<{ collections: Collection[] }>(`/collections?workspace=${workspaceId}`);
      return response.collections || [];
    } catch {
      return [];
    }
  }

  /**
   * Get collection details
   */
  async getCollectionDetails(collectionUid: string): Promise<CollectionDetails | null> {
    try {
      const response = await this.http.get<{ collection: CollectionDetails }>(`/collections/${collectionUid}`);
      return response.collection;
    } catch {
      return null;
    }
  }

  /**
   * Fork a collection
   */
  async forkCollection(
    collectionUid: string,
    label: string,
    targetWorkspaceId: string
  ): Promise<ForkResult> {
    try {
      const response = await this.http.post<{ collection: Collection }>(
        `/collections/fork/${collectionUid}?workspace=${targetWorkspaceId}`,
        { label }
      );
      return { success: true, collection: response.collection };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Delete a collection
   */
  async deleteCollection(collectionUid: string): Promise<boolean> {
    try {
      await this.http.delete(`/collections/${collectionUid}`);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Update a collection's variables via partial update
   * PATCH /collections/{collectionId}
   */
  async patchCollectionVariables(
    collectionUid: string,
    variables: Array<{ key: string; value: string; [k: string]: unknown }>
  ): Promise<{ success: boolean; error?: string }> {
    try {
      await this.http.patch(`/collections/${collectionUid}`, {
        collection: { variable: variables },
      });
      return { success: true };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }

  // =========================================================================
  // ENVIRONMENTS
  // =========================================================================

  /**
   * Get all environments in a workspace
   */
  async getEnvironments(workspaceId: string): Promise<Environment[]> {
    try {
      const response = await this.http.get<{ environments: Environment[] }>(`/environments?workspace=${workspaceId}`);
      return response.environments || [];
    } catch {
      return [];
    }
  }

  /**
   * Get environment details with variables
   */
  async getEnvironmentDetails(environmentUid: string): Promise<EnvironmentDetails | null> {
    try {
      const response = await this.http.get<{ environment: EnvironmentDetails }>(`/environments/${environmentUid}`);
      return response.environment;
    } catch {
      return null;
    }
  }

  /**
   * Create environment
   */
  async createEnvironment(
    name: string,
    values: EnvironmentVariable[],
    workspaceId: string
  ): Promise<CreateEnvironmentResult> {
    try {
      const response = await this.http.post<{ environment: Environment }>(`/environments?workspace=${workspaceId}`, {
        environment: { name, values },
      });
      return { success: true, environment: response.environment };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Update environment (full replace)
   */
  async updateEnvironment(
    environmentUid: string,
    name: string,
    values: EnvironmentVariable[]
  ): Promise<UpdateEnvironmentResult> {
    try {
      const response = await this.http.put<{ environment: Environment }>(`/environments/${environmentUid}`, {
        environment: { name, values },
      });
      return { success: true, environment: response.environment };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Patch environment (partial update)
   */
  async patchEnvironment(
    environmentUid: string,
    operations: PatchEnvironmentOperation[]
  ): Promise<UpdateEnvironmentResult> {
    try {
      const response = await this.http.patch<{ environment: Environment }>(`/environments/${environmentUid}`, operations);
      return { success: true, environment: response.environment };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Delete environment
   */
  async deleteEnvironment(environmentUid: string): Promise<boolean> {
    try {
      await this.http.delete(`/environments/${environmentUid}`);
      return true;
    } catch {
      return false;
    }
  }

  // =========================================================================
  // MOCK SERVERS
  // =========================================================================

  /**
   * Get all mocks in a workspace
   */
  async getMocks(workspaceId: string): Promise<MockServer[]> {
    try {
      const response = await this.http.get<{ mocks: MockServer[] }>(`/mocks?workspace=${workspaceId}`);
      return response.mocks || [];
    } catch {
      return [];
    }
  }

  /**
   * Get mock details
   */
  async getMockDetails(mockId: string): Promise<MockServer | null> {
    try {
      const response = await this.http.get<{ mock: MockServer }>(`/mocks/${mockId}`);
      return response.mock;
    } catch {
      return null;
    }
  }

  /**
   * Create mock server
   */
  async createMock(request: CreateMockRequest): Promise<CreateMockResult> {
    try {
      const mockConfig: Record<string, unknown> = {
        name: request.name,
        collection: request.collection,
        private: request.isPrivate ?? false,
      };

      if (request.environment) {
        mockConfig.environment = request.environment;
      }

      if (request.config) {
        mockConfig.config = request.config;
      }

      const response = await this.http.post<{ mock: MockServer }>(`/mocks?workspace=${request.workspaceId}`, {
        mock: mockConfig,
      });
      return { success: true, mock: response.mock };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Delete mock server
   */
  async deleteMock(mockId: string): Promise<boolean> {
    try {
      await this.http.delete(`/mocks/${mockId}`);
      return true;
    } catch {
      return false;
    }
  }

  // =========================================================================
  // SPECS
  // =========================================================================

  /**
   * Get all specs in a workspace
   */
  async getSpecs(workspaceId: string): Promise<Spec[]> {
    try {
      const response = await this.http.get<{ specs: Spec[] }>(`/specs?workspaceId=${workspaceId}`);
      return response.specs || [];
    } catch {
      return [];
    }
  }

  /**
   * Get spec details
   */
  async getSpecDetails(specId: string): Promise<Spec | null> {
    try {
      const response = await this.http.get<Spec>(`/specs/${specId}`);
      return response;
    } catch {
      return null;
    }
  }

  /**
   * Get all files in a spec
   */
  async getSpecFiles(specId: string): Promise<SpecFile[]> {
    try {
      const response = await this.http.get<{ files: SpecFile[] }>(`/specs/${specId}/files`);
      return response.files || [];
    } catch {
      return [];
    }
  }

  /**
   * Get a specific spec file with content
   */
  async getSpecFile(specId: string, filePath: string): Promise<SpecFileWithContent | null> {
    try {
      const encodedPath = encodeURIComponent(filePath);
      const response = await this.http.get<SpecFileWithContent>(`/specs/${specId}/files/${encodedPath}`);
      return response;
    } catch {
      return null;
    }
  }

  /**
   * Create a spec with files
   */
  async createSpec(
    workspaceId: string,
    name: string,
    type: SpecType,
    files: CreateSpecFile[]
  ): Promise<CreateSpecResult> {
    try {
      const response = await this.http.post<Spec>(`/specs?workspaceId=${workspaceId}`, {
        name,
        type,
        files,
      });
      return { success: true, spec: response };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  /**
   * Delete a spec
   */
  async deleteSpec(specId: string): Promise<DeleteSpecResult> {
    try {
      await this.http.delete(`/specs/${specId}`);
      return { success: true };
    } catch (error) {
      return { success: false, error: getErrorMessage(error) };
    }
  }

  // =========================================================================
  // PARTNER INVITATIONS
  // =========================================================================

  /**
   * Invite partner to workspace
   */
  async invitePartner(
    workspaceId: string,
    email: string,
    roleId: string = WorkspaceRoleId.PartnerEditorAndLead
  ): Promise<InvitePartnerResult> {
    try {
      const response = await this.http.post<{
        results: Array<{
          email: string;
          status: string;
          invitationLink?: string;
          userId?: string;
        }>;
        roleDisplayName?: string;
      }>('/invitations', {
        action: 'invite_partner',
        targetEntity: 'workspace',
        targetEntityId: workspaceId,
        roleId,
        target: { emails: [email] },
      });

      const result = response.results?.[0];
      return {
        success: true,
        email: result?.email || email,
        status: result?.status as InvitePartnerResult['status'],
        invitationLink: result?.invitationLink ?? null,
        userId: result?.userId ?? null,
        roleDisplayName: response.roleDisplayName,
      };
    } catch (error) {
      return { success: false, email, error: getErrorMessage(error) };
    }
  }

  /**
   * Remove partner from workspace
   */
  async removePartner(workspaceId: string, userId: string): Promise<RemovePartnerResult> {
    try {
      const response = await this.http.post<{
        results: Array<{ userId: string; status: string }>;
      }>('/invitations', {
        action: 'remove_partner',
        targetEntity: 'workspace',
        targetEntityId: workspaceId,
        target: { userIds: [userId] },
      });

      const result = response.results?.[0];
      return {
        success: true,
        userId: result?.userId || userId,
        status: result?.status,
      };
    } catch (error) {
      return { success: false, userId, error: getErrorMessage(error) };
    }
  }

  /**
   * Remove partner from team
   */
  async removePartnerFromTeam(teamId: string, userId: string): Promise<RemovePartnerResult> {
    try {
      const response = await this.http.post<{
        results: Array<{ userId: string; status: string }>;
      }>('/invitations', {
        action: 'remove_partner',
        targetEntity: 'team',
        targetEntityId: teamId,
        target: { userIds: [userId] },
      });

      const result = response.results?.[0];
      return {
        success: true,
        userId: result?.userId || userId,
        status: result?.status,
      };
    } catch (error) {
      return { success: false, userId, error: getErrorMessage(error) };
    }
  }
}
